LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsData.cs

```diff
@@ -40,5 +34,106 @@ public static class LnsData
+               */
+
+        internal static class RadioMetadataProperties
+        {
+            public static readonly IJsonProperty<DataRate> DataRate =
+                JsonReader.Property("DR", from n in JsonReader.Byte() select new DataRate(n));
+
+            public static readonly IJsonProperty<Hertz> Freq =
+                JsonReader.Property("Freq", from n in JsonReader.UInt32() select new Hertz(n));
+
+            public static readonly IJsonProperty<RadioMetadataUpInfo> UpInfo =
+                JsonReader.Property("upinfo",
+                                    JsonReader.Object(JsonReader.Property("rctx", JsonReader.UInt32()),
+                                                      JsonReader.Property("xtime", JsonReader.UInt64()),
+                                                      JsonReader.Property("gpstime", JsonReader.UInt32()),
+                                                      JsonReader.Property("rssi", JsonReader.Double()),
+                                                      JsonReader.Property("snr", JsonReader.Single()),
+                                                      (Rctx, Xtime, GpsTime, Rssi, Snr) => new RadioMetadataUpInfo(Rctx, Xtime, GpsTime, Rssi, Snr)));
+        }
+
+        /*
+            {
+              "msgtype"   : "updf",
+              "MHdr"      : UINT,
+              "DevAddr"   : INT32,
+              "FCtrl"     : UINT,
+              "FCnt",     : UINT,
+              "FOpts"     : "HEX",
+              "FPort"     : INT(-1..255),
+              "FRMPayload": "HEX",
+              "MIC"       : INT32,
+              ..
+              RADIOMETADATA
+            }
+         */
+
+        internal static readonly IJsonReader<UpstreamDataFrame> UpstreamDataFrameReader =
+            JsonReader.Object(MessageTypeProperty(LnsMessageType.UplinkDataFrame),
+                              JsonReader.Property("MHdr", JsonReader.Byte()),
+                              JsonReader.Property("DevAddr", JsonReader.UInt32()),
+                              JsonReader.Property("FCtrl", JsonReader.Byte()),
+                              JsonReader.Property("FCnt", JsonReader.UInt16()),
+                              JsonReader.Property("FOpts", JsonReader.String()),
+                              JsonReader.Property("FPort", JsonReader.Byte()),
+                              JsonReader.Property("FRMPayload", JsonReader.String()),
+                              JsonReader.Property("MIC", JsonReader.Int32()),
+                              RadioMetadataProperties.DataRate,
+                              RadioMetadataProperties.Freq,
+                              RadioMetadataProperties.UpInfo,
+                              (_, MHdr, DevAddr, FCtrl, FCnt, FOpts, FPort, FRMPayload, MIC, dr, freq, upInfo) =>
+                                new UpstreamDataFrame(new MacHeader(MHdr), new DevAddr(DevAddr), new FrameControl(FCtrl), FCnt, FOpts, new FramePort(FPort), FRMPayload, new Mic((uint)MIC),
```

Comment:
Please use camel casing for arguments per .NET conventions:

```suggestion
                              (_, mhdr, devAddr, ctrl, cnt, opts, port, payload, mic, dr, freq, upInfo) =>
                                new UpstreamDataFrame(new MacHeader(mhdr), new DevAddr(devAddr), new FrameControl(ctrl), cnt, opts, new FramePort(port), payload, new Mic((uint)mic),
```